// @datastax/astra-db-ts/dist/cjs/lib/api/timeouts/cfg-handler.js
"use strict";
// Copyright Datastax, Inc
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeoutCfgHandler = void 0;
const opts_handlers_js_1 = require("../../../lib/opts-handlers.js");
const decoders_1 = require("decoders");
const monoid = opts_handlers_js_1.monoids.object({
    requestTimeoutMs: opts_handlers_js_1.monoids.optional(),
    generalMethodTimeoutMs: opts_handlers_js_1.monoids.optional(),
    collectionAdminTimeoutMs: opts_handlers_js_1.monoids.optional(),
    tableAdminTimeoutMs: opts_handlers_js_1.monoids.optional(),
    databaseAdminTimeoutMs: opts_handlers_js_1.monoids.optional(),
    keyspaceAdminTimeoutMs: opts_handlers_js_1.monoids.optional(),
});
const decoder = (0, decoders_1.nullish)((0, decoders_1.exact)({
    requestTimeoutMs: (0, decoders_1.optional)(decoders_1.positiveNumber),
    generalMethodTimeoutMs: (0, decoders_1.optional)(decoders_1.positiveNumber),
    collectionAdminTimeoutMs: (0, decoders_1.optional)(decoders_1.positiveNumber),
    tableAdminTimeoutMs: (0, decoders_1.optional)(decoders_1.positiveNumber),
    databaseAdminTimeoutMs: (0, decoders_1.optional)(decoders_1.positiveNumber),
    keyspaceAdminTimeoutMs: (0, decoders_1.optional)(decoders_1.positiveNumber),
}), {});
exports.TimeoutCfgHandler = new opts_handlers_js_1.MonoidalOptionsHandler(decoder, monoid);
