// @datastax/astra-db-ts/dist/cjs/administration/db-admin.js
"use strict";
// Copyright Datastax, Inc
// SPDX-License-Identifier: Apache-2.0
// noinspection ExceptionCaughtLocallyJS
Object.defineProperty(exports, "__esModule", { value: true });
exports.DbAdmin = void 0;
const index_js_1 = require("../lib/index.js");
class DbAdmin extends index_js_1.HierarchicalLogger {
        async findEmbeddingProviders(options) {
        const httpClient = this._getDataAPIHttpClient();
        const resp = await httpClient.executeCommand({ findEmbeddingProviders: {} }, {
            timeoutManager: httpClient.tm.single('databaseAdminTimeoutMs', options),
            methodName: 'dbAdmin.findEmbeddingProviders',
            keyspace: null,
        });
        return resp.status;
    }
            async findRerankingProviders(options) {
        const httpClient = this._getDataAPIHttpClient();
        const resp = await httpClient.executeCommand({ findRerankingProviders: {} }, {
            timeoutManager: httpClient.tm.single('databaseAdminTimeoutMs', options),
            methodName: 'dbAdmin.findRerankingProviders',
            keyspace: null,
        });
        return resp.status;
    }
}
exports.DbAdmin = DbAdmin;
