// @datastax/astra-db-ts/dist/esm/administration/db-admin.js
// Copyright Datastax, Inc
// SPDX-License-Identifier: Apache-2.0
// noinspection ExceptionCaughtLocallyJS
import { HierarchicalLogger } from '../lib/index.js';
export class DbAdmin extends HierarchicalLogger {
        async findEmbeddingProviders(options) {
        const httpClient = this._getDataAPIHttpClient();
        const resp = await httpClient.executeCommand({ findEmbeddingProviders: {} }, {
            timeoutManager: httpClient.tm.single('databaseAdminTimeoutMs', options),
            methodName: 'dbAdmin.findEmbeddingProviders',
            keyspace: null,
        });
        return resp.status;
    }
            async findRerankingProviders(options) {
        const httpClient = this._getDataAPIHttpClient();
        const resp = await httpClient.executeCommand({ findRerankingProviders: {} }, {
            timeoutManager: httpClient.tm.single('databaseAdminTimeoutMs', options),
            methodName: 'dbAdmin.findRerankingProviders',
            keyspace: null,
        });
        return resp.status;
    }
}
