// @datastax/astra-db-ts/dist/esm/lib/logging/hierarchical-logger.js
// Copyright Datastax, Inc
// SPDX-License-Identifier: Apache-2.0
import { InternalLogger } from '../../lib/logging/internal-logger.js';
export class HierarchicalLogger {
        constructor(parent, config) {
                Object.defineProperty(this, "internal", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.internal = new InternalLogger(config, parent?.internal, console);
    }
    updateLoggingConfig(config) {
        this.internal.updateLoggingConfig(InternalLogger.cfg.parse(config));
    }
        on(eventName, listener) {
        this.internal.on(eventName, listener);
        return () => {
            this.off(eventName, listener);
        };
    }
        off(eventName, listener) {
        return this.internal.off(eventName, listener);
    }
        once(eventName, listener) {
        const off = this.on(eventName, (event) => {
            off();
            listener(event);
        });
        return off;
    }
        removeAllListeners(eventName) {
        return this.internal.removeAllListeners(eventName);
    }
}
