// @datastax/astra-db-ts/dist/esm/lib/logging/internal-logger.js
// Copyright Datastax, Inc
// SPDX-License-Identifier: Apache-2.0
// noinspection DuplicatedCode
import { EmptyInternalLoggingConfig, EventConstructors, PrintLoggingOutputs } from '../../lib/logging/constants.js';
import { buildOutputsMap } from '../../lib/logging/util.js';
import { PropagationState } from '../../lib/index.js';
import { LoggingCfgHandler } from '../../lib/logging/cfg-handler.js';
import * as uuid from 'uuid';
export class InternalLogger {
    constructor(config, parent, console) {
        Object.defineProperty(this, "commandStarted", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "commandFailed", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "commandWarnings", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "commandSucceeded", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "adminCommandFailed", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "adminCommandStarted", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "adminCommandPolling", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "adminCommandWarnings", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "adminCommandSucceeded", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "_someCommandEventEnabled", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
        Object.defineProperty(this, "_someAdminCommandEventEnabled", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
        Object.defineProperty(this, "_config", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "_listeners", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: Object.create(null)
        });
        Object.defineProperty(this, "_parent", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "_console", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this._parent = parent;
        this._console = console;
        this._config = this._buildInternalConfig(config);
    }
    generateCommandRequestId() {
        return this._someCommandEventEnabled ? uuid.v4() : '';
    }
    generateAdminCommandRequestId() {
        return this._someAdminCommandEventEnabled ? uuid.v4() : '';
    }
    updateLoggingConfig(config) {
        this._config = this._updateInternalConfig(this._config, config);
    }
    on(eventName, listener) {
        if (!this._listeners[eventName]) {
            this._listeners[eventName] = [];
        }
        this._listeners[eventName].push(listener);
    }
    off(eventName, listener) {
        if (!this._listeners[eventName]) {
            return;
        }
        const index = this._listeners[eventName].indexOf(listener);
        if (index !== -1) {
            this._listeners[eventName].splice(index, 1);
        }
        if (this._listeners[eventName].length === 0) {
            delete this._listeners[eventName];
        }
    }
    removeAllListeners(eventName) {
        if (eventName) {
            delete this._listeners[eventName];
        }
        else {
            for (const key in this._listeners) {
                delete this._listeners[key];
            }
        }
    }
    emit(eventName, event) {
        if (this._listeners[eventName]) {
            for (const listener of this._listeners[eventName]) {
                try {
                    listener(event);
                }
                catch (_e) {
                    // Silently ignore errors
                }
                if (event._propagationState === PropagationState.StopImmediate) {
                    return;
                }
            }
        }
        if (this._parent && event._propagationState !== PropagationState.Stop) {
            this._parent.emit(eventName, event);
        }
    }
    _buildInternalConfig(config) {
        return this._updateInternalConfig(EmptyInternalLoggingConfig, config);
    }
    _updateInternalConfig(base, config) {
        const newConfig = { ...base };
        for (const layer of config.layers) {
            for (const event of layer.events) {
                newConfig[event] = buildOutputsMap(layer.emits);
                const activeOutputs = PrintLoggingOutputs.filter(key => newConfig[event]?.[key]);
                if (activeOutputs.length > 1) {
                    throw new Error(`Nonsensical logging configuration; conflicting outputs '${activeOutputs}' set for '${event}'`);
                }
            }
        }
        this._buildLoggingFunctions(newConfig);
        return newConfig;
    }
    _buildLoggingFunctions(config) {
        this._someCommandEventEnabled = false;
        this._someAdminCommandEventEnabled = false;
        for (const [_eventName, outputs] of Object.entries(config)) {
            const eventName = _eventName;
            if (!outputs) {
                this[eventName] = undefined;
                continue;
            }
            const log = this._mkLogFn(outputs);
            if (eventName.startsWith('admin')) {
                this._someAdminCommandEventEnabled = true;
            }
            else {
                this._someCommandEventEnabled = true;
            }
            this[eventName] = (...args) => {
                const event = new EventConstructors[eventName](...args);
                outputs.event && this.emit(eventName, event);
                log?.(event);
            };
        }
    }
    _mkLogFn(outputs) {
        switch (true) {
            case outputs.stdout:
                return (event) => this._console.log(event.format());
            case outputs.stderr:
                return (event) => this._console.error(event.format());
            case outputs['stdout:verbose']:
                return (event) => this._console.log(event.formatVerbose());
            case outputs['stderr:verbose']:
                return (event) => this._console.error(event.formatVerbose());
        }
    }
}
Object.defineProperty(InternalLogger, "cfg", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: LoggingCfgHandler
});
