// @datastax/astra-db-ts/dist/cjs/documents/collections/index.js
"use strict";
// Copyright Datastax, Inc
// SPDX-License-Identifier: Apache-2.0
// noinspection DuplicatedCode
Object.defineProperty(exports, "__esModule", { value: true });
exports.NumCoercionError = exports.$DeserializeForCollection = exports.$SerializeForCollection = exports.CollectionCodecs = exports.Collection = exports.CollectionFindAndRerankCursor = void 0;
const tslib_1 = require("tslib");
tslib_1.__exportStar(require("./cursors/find-cursor.js"), exports);
var rerank_cursor_js_1 = require("./cursors/rerank-cursor.js");
Object.defineProperty(exports, "CollectionFindAndRerankCursor", { enumerable: true, get: function () { return rerank_cursor_js_1.CollectionFindAndRerankCursor; } });
var collection_js_1 = require("./collection.js");
Object.defineProperty(exports, "Collection", { enumerable: true, get: function () { return collection_js_1.Collection; } });
tslib_1.__exportStar(require("./ser-des/codecs.js"), exports);
var codecs_js_1 = require("./ser-des/codecs.js"); // 2nd export necessary for some reason; otherwise CollectionCodecs isn't available in CJS
Object.defineProperty(exports, "CollectionCodecs", { enumerable: true, get: function () { return codecs_js_1.CollectionCodecs; } });
var constants_js_1 = require("./ser-des/constants.js");
Object.defineProperty(exports, "$SerializeForCollection", { enumerable: true, get: function () { return constants_js_1.$SerializeForCollection; } });
Object.defineProperty(exports, "$DeserializeForCollection", { enumerable: true, get: function () { return constants_js_1.$DeserializeForCollection; } });
var big_nums_js_1 = require("./ser-des/big-nums.js");
Object.defineProperty(exports, "NumCoercionError", { enumerable: true, get: function () { return big_nums_js_1.NumCoercionError; } });
