// @datastax/astra-db-ts/dist/esm/documents/cursors/find-cursor.js
// Copyright Datastax, Inc
// SPDX-License-Identifier: Apache-2.0
import { vector } from '../../documents/datatypes/vector.js';
import { AbstractCursor } from '../../documents/cursors/abstract-cursor.js';
import { $CustomInspect } from '../../lib/constants.js';
import { SerDesTarget } from '../../lib/api/ser-des/ctx.js';
import { QueryState } from '../../lib/utils.js';
import { buildFLCFilter, buildFLCMap, buildFLCOption, buildFLCPreMapOption, buildFLCSort, cloneFLC, } from '../../documents/cursors/common.js';
export class FindCursor extends AbstractCursor {
        constructor(parent, serdes, filter, options, mapping) {
        super(options ?? {}, mapping);
                Object.defineProperty(this, "_httpClient", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
                Object.defineProperty(this, "_serdes", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
                Object.defineProperty(this, "_parent", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
                Object.defineProperty(this, "_filter", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
                Object.defineProperty(this, "_sortVector", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new QueryState()
        });
        this._parent = parent;
        this._httpClient = parent._httpClient;
        this._serdes = serdes;
        this._filter = filter;
    }
        [$CustomInspect]() {
        return `${this.constructor.name}(source="${this._parent.keyspace}.${this._parent.name}",state="${this._state}",consumed=${this.consumed()},buffered=${this.buffered()})`;
    }
        filter(filter) {
        return buildFLCFilter(this, filter);
    }
        sort(sort) {
        return buildFLCSort(this, sort);
    }
        limit(limit) {
        return buildFLCOption(this, 'limit', limit || undefined);
    }
        skip(skip) {
        return buildFLCOption(this, 'skip', skip);
    }
        includeSortVector(includeSortVector) {
        return buildFLCOption(this, 'includeSortVector', includeSortVector ?? true);
    }
        project(projection) {
        return buildFLCPreMapOption(this, 'projection', structuredClone(projection));
    }
        includeSimilarity(includeSimilarity) {
        return buildFLCPreMapOption(this, 'includeSimilarity', includeSimilarity ?? true);
    }
        map(map) {
        return buildFLCMap(this, map);
    }
        async getSortVector() {
        if (this._sortVector.state === QueryState.Unattempted && this._options.includeSortVector) {
            const reset2idle = this._state === 'idle';
            await this._next(true, '.getSortVector');
            if (reset2idle) {
                this._state = 'idle';
            }
        }
        return this._sortVector.unwrap();
    }
        clone() {
        return cloneFLC(this, this._filter, this._options, this._mapping);
    }
        async _nextPage(extra, tm) {
        const command = {
            find: {
                filter: this._filter[0],
                projection: this._options.projection,
                sort: this._options.sort,
                options: {
                    includeSimilarity: this._options.includeSimilarity,
                    includeSortVector: this._options.includeSortVector,
                    limit: this._options.limit,
                    skip: this._options.skip,
                    pageState: this._nextPageState.unwrap(),
                },
            },
        };
        const raw = await this._httpClient.executeCommand(command, {
            timeoutManager: tm ?? this._httpClient.tm.single('generalMethodTimeoutMs', this._options),
            bigNumsPresent: this._filter[1],
            extraLogInfo: extra,
        });
        this._nextPageState.swap(raw.data?.nextPageState);
                const buffer = raw.data?.documents ?? [];
        for (let i = 0, n = buffer.length; i < n; i++) {
            buffer[i] = this._serdes.deserialize(buffer[i], raw, SerDesTarget.Record);
        }
        const sortVector = raw.status?.sortVector;
        this._sortVector.swap(sortVector ? vector(sortVector) : sortVector);
        this._options.includeSortVector = false;
        return buffer;
    }
        _tm() {
        return this._httpClient.tm;
    }
}
